timer/src/main/java/com/toggl/timer/project/ui/ColorHueSaturationPicker.kt

```diff
@@ -0,0 +1,130 @@
+package com.toggl.timer.project.ui
+
+import android.content.Context
+import android.graphics.Canvas
+import android.graphics.Color
+import android.graphics.LinearGradient
+import android.graphics.Paint
+import android.graphics.RectF
+import android.graphics.Shader
+import android.util.AttributeSet
+import android.view.MotionEvent
+import android.view.View
+import androidx.core.graphics.ColorUtils
+import com.toggl.timer.R
+import kotlin.math.abs
+import kotlin.properties.Delegates
+
+class ColorHueSaturationPicker @JvmOverloads constructor(
+    context: Context,
+    attrs: AttributeSet? = null,
+    defStyleAttr: Int = 0
+) : View(context, attrs, defStyleAttr) {
+
+    private val brightnessGradientColors = intArrayOf(Color.TRANSPARENT, Color.WHITE)
+    private val colorGradientColors = intArrayOf(
+        Color.RED, Color.YELLOW, Color.GREEN, Color.CYAN, Color.BLUE, Color.MAGENTA, Color.RED
+    )
+
+    private var circleRadius: Float = 0f
+    private var circleDiameter: Float = 0f
+    private val circleFillPaint = Paint().apply {
+        style = Paint.Style.FILL
+        flags = Paint.ANTI_ALIAS_FLAG
+    }
+    private val circleStrokePaint: Paint
+    private val circleColor = Color.WHITE
+
+    private val brightnessPaint = Paint().apply {
+        style = Paint.Style.FILL
+        flags = Paint.ANTI_ALIAS_FLAG
+    }
+    private val opacityBackgroundPaint = Paint().apply {
+        style = Paint.Style.FILL
+        flags = Paint.ANTI_ALIAS_FLAG
+    }
+    private val gradientBackgroundPaint = Paint().apply {
+        style = Paint.Style.FILL
+        flags = Paint.ANTI_ALIAS_FLAG
+    }
+
+    var hue: Float = 0f
+
+    var saturation: Float = 0f
+
+    var value: Float by Delegates.observable(1f) { _, oldVal, newValue ->
+        if (newValue != oldVal) {
+            val opacity = complement(value) * 255
+            opacityBackgroundPaint.color = ColorUtils.setAlphaComponent(Color.BLACK, opacity.toInt())
+            invalidate()
+        }
+    }
+
+    var onHueSaturationChangedListener: ((Float, Float) -> Unit)? = null
+
+    init {
+        circleDiameter = resources.getDimension(R.dimen.color_hue_saturation_picker_circle_diameter)
+        circleRadius = circleDiameter / 2
+
+        circleStrokePaint = Paint().apply {
+            style = Paint.Style.STROKE
+            color = circleColor
+            strokeCap = Paint.Cap.ROUND
+            strokeWidth = resources.getDimension(R.dimen.color_hue_saturation_picker_stroke_width)
+            flags = Paint.ANTI_ALIAS_FLAG
+        }
+    }
+
+    override fun onTouchEvent(e: MotionEvent): Boolean =
+        when (e.action) {
+            MotionEvent.ACTION_MOVE,
+            MotionEvent.ACTION_DOWN -> {
+                updateLocationTo(e.x, e.y)
+                true
+            }
+            else -> super.onTouchEvent(e)
```

Comment:
```suggestion
            MotionEvent.ACTION_DOWN -> {
                parent.requestDisallowInterceptTouchEvent(true)
                updateLocationTo(e.x, e.y)
                true
            }
            MotionEvent.ACTION_UP -> {
                parent.requestDisallowInterceptTouchEvent(false)
                true
            }
            else -> super.onTouchEvent(e)
```

Seems to be enough to prevent the bottomsheet parent from consuming touches while this view is handling them.
I think we can use this idea on the wheel as well 🤔

You might need to handle `ACTION_CANCEL` as well to give the control back to the parent.

Reply:
Looks good works well 👍